src/base/bittorrent/sessionimpl.cpp

```diff
@@ -2892,26 +2897,44 @@ void SessionImpl::saveResumeData()
         ++m_numResumeData;
     }
 
+    QElapsedTimer timer;
+    timer.start();
+
     while (m_numResumeData > 0)
     {
-        const std::vector<lt::alert *> alerts = getPendingAlerts(lt::seconds {30});
-        if (alerts.empty())
-        {
-            LogMsg(tr("Aborted saving resume data. Number of outstanding torrents: %1").arg(QString::number(m_numResumeData))
-                , Log::CRITICAL);
-            break;
-        }
+        const lt::seconds waitTime {30};
+        const std::vector<lt::alert *> alerts = getPendingAlerts(waitTime);
 
+        bool hasWantedAlert = false;
         for (const lt::alert *a : alerts)
         {
             switch (a->type())
             {
             case lt::save_resume_data_failed_alert::alert_type:
+                hasWantedAlert = true;
                 --m_numResumeData;
                 break;
             case lt::save_resume_data_alert::alert_type:
+                hasWantedAlert = true;
                 dispatchTorrentAlert(static_cast<const lt::torrent_alert *>(a));
                 break;
+            default:
+                // skip over other alert types
+                break;
+            }
+        }
+
+        if (hasWantedAlert)
+        {
+            timer.start();
+        }
+        else
+        {
+            if (timer.hasExpired(lt::total_milliseconds(waitTime)))
```

Comment:
I would merge it with `else` above (i.e. write it as `else if`).

Reply:
Done.